src/main/kotlin/org/jetbrains/kotlin/jupyter/protocol.kt

```diff
@@ -169,12 +173,50 @@ fun JupyterConnection.Socket.shellMessagesHandler(msg: Message, repl: ReplForJup
     }
 }
 
-class CapturingOutputStream(val stdout: PrintStream, val captureOutput: Boolean) : OutputStream() {
+class CapturingOutputStream(private val stdout: PrintStream,
+                            private val conf: OutputConfig,
+                            private val captureOutput: Boolean,
+                            val onCaptured: (String) -> Unit) : OutputStream() {
     val capturedOutput = ByteArrayOutputStream()
+    private var time = System.currentTimeMillis()
+    private var overallOutputSize = 0
+    private var newlineFound = false
+
+    private fun shouldSend(b: Int): Boolean {
+        val c = b.toChar()
+        newlineFound = newlineFound || c == '\n' || c == '\r'
+        if (newlineFound && capturedOutput.size() >= conf.captureNewlineBufferSize)
+            return true
+        if (capturedOutput.size() >= conf.captureBufferMaxSize)
+            return true
+
+        val currentTime = System.currentTimeMillis()
+        if (currentTime - time >= conf.captureBufferTimeLimitMs) {
+            time = currentTime
+            return true
+        }
+        return false
+    }
 
     override fun write(b: Int) {
+        ++overallOutputSize
         stdout.write(b)
-        if (captureOutput) capturedOutput.write(b)
+
+        if (captureOutput && overallOutputSize <= conf.cellOutputMaxSize) {
+            capturedOutput.write(b)
```

Comment:
This causes sending first character of next line together with previous line (if time since previous line output > timeLimit)

Reply:
Yes, it was what I expected it test, and t seemed to me overcomplicated to make timer. But I think it's ok to make things good from the beginning, so i've implemented it.